rasa/engine/caching.py

```diff
@@ -0,0 +1,450 @@
+                self.CacheEntry.last_used.asc()
+            )
+            oldest_cache_item = (
+                session.execute(query_for_least_recently_used_entry).scalars().first()
+            )
+
+            if not oldest_cache_item:
+                self._purge_cache_dir_content()
+                return
+
+            self._delete_cached_result(oldest_cache_item)
+            delete_query = sa.delete(self.CacheEntry).where(
+                self.CacheEntry.fingerprint_key == oldest_cache_item.fingerprint_key
+            )
+            session.execute(delete_query)
+
+            logger.debug(
+                f"Deleted item with fingerprint "
+                f"'{oldest_cache_item.fingerprint_key}' to free space."
+            )
+
+    def _purge_cache_dir_content(self) -> None:
+        for item in self._cache_location.glob("*"):
+            if item.name == self._cache_database_name:
+                continue
+
+            if item.is_dir():
+                shutil.rmtree(item)
+            else:
+                item.unlink()
+
+    def get_cached_output_fingerprint(self, fingerprint_key: Text) -> Optional[Text]:
+        """Returns cached output fingerprint (see parent class for full docstring)."""
+        with self._sessionmaker.begin() as session:
+            query = sa.select(self.CacheEntry).filter_by(
+                fingerprint_key=fingerprint_key
+            )
+            match = session.execute(query).scalars().first()
+
+            if match:
+                # This result was used during a fingerprint run.
+                match.last_used = datetime.utcnow()
+                return match.output_fingerprint_key
+
+            return None
+
+    def get_cached_result(
+        self, output_fingerprint_key: Text, node_name: Text, model_storage: ModelStorage
+    ) -> Optional[Cacheable]:
+        """Returns a potentially cached output (see parent class for full docstring)."""
+        result_location, result_type = self._get_cached_result(output_fingerprint_key)
+
+        if not result_location:
+            logger.debug(f"No cached output found for '{output_fingerprint_key}'")
+            return None
+
+        path_to_cached = Path(result_location)
+        if not path_to_cached.is_dir():
+            logger.debug(
+                f"Cached output for '{output_fingerprint_key}' can't be found on disk."
+            )
+            return None
+
+        return self._load_from_cache(
+            result_location, result_type, node_name, model_storage
+        )
+
+    def _get_cached_result(
+        self, output_fingerprint_key: Text
+    ) -> Tuple[Optional[Path], Optional[Text]]:
+        with self._sessionmaker.begin() as session:
+            query = sa.select(
+                self.CacheEntry.result_location, self.CacheEntry.result_type
+            ).where(
+                self.CacheEntry.output_fingerprint_key == output_fingerprint_key,
+                self.CacheEntry.result_location != sa.null(),
```

Comment:
More of a question: why do we need `self.CacheEntry.result_location != sa.null()` condition? I understand that when the result location is empty then there is basically no cached output but why it's the case and in which situations can it happen?

Reply:
Every `GraphComponent` output has to be `Fingerprintable` but not every `GraphComponent` output is `Cacheable`. This means that we sometimes only store the fingerprints of the output but not the cached output. This is e.g. allows cases like this

```
# Train
A (not cachable, no target node) -> B (cacheable, target node)

# Fingerprint run
A (fingerprint didn't change)  -> B (fingerprint didn't change)

# This leads to the "train graph" only consisting of a cached output of B 
# although we don't have a cached version of A
B (using cached version)

```

Does that make sense?

Reply:
Yes, makes sense! Thanks for the explanation